From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: IBM Codestream linux-2.6.16 october2005, patch 02-12
Patch-Mainline: 2.6.17-rc4

  Description: dasd: BIODASDENABLE ioctl never returns on unformatted device.
  Symptom:     A program calling BIODASDENABLE on an unformatted device never
               returns.
  Problem:     The DASD state- machine is not designed to enable an unformatted
               device, since 'unformatted' is a final state.
               Problem is, that this final state is not detected so far and the
               state-machine is waiting forever.
               Note: To get such a device online it has to be re-analyzed.
               This means that the device needs to be disabled prior to
               re-enablement.
  Solution:    Handle 'unformatted' as a final state and return.
  Problem-ID:  -
  Archs:       s390-31, s390-64

Acked-by: Hannes Reinecke <hare@suse.de>

--- linux-2.5/drivers/s390/block/dasd.c	10 Apr 2006 12:32:01 -0000	1.181.2.3
+++ linux-2.5/drivers/s390/block/dasd.c	19 Apr 2006 13:42:26 -0000	1.181.2.4
@@ -315,6 +315,11 @@
 		rc = dasd_state_basic_to_ready(device);
 
 	if (!rc &&
+	    device->state == DASD_STATE_UNFMT &&
+	    device->target > DASD_STATE_UNFMT)
+		rc = -EPERM;
+	
+	if (!rc &&
 	    device->state == DASD_STATE_READY &&
 	    device->target >= DASD_STATE_ONLINE)
 		rc = dasd_state_ready_to_online(device);

